pastas/recharge.py

```diff
@@ -146,116 +167,347 @@ class FlexModel(RechargeBase):
 
         R = K_s \\left( \\frac{S}{S_u}\\right) ^\\gamma
 
-    For a detailed description of the recharge model and parameters we refer
-    to Collenteur et al. (in review).
+    If snow=True, a snow reservoir is added on top. For a detailed
+    description of the degree-day snow model and parameters we refer to
+    [kavetski_2007]_. The water balance for the snow reservoir is written as:
+
+    .. math::
+
+        \\frac{dSs}{dt} = Ps - M
+
+    Note that the preferred unit of the precipitation and evaporation is
+    mm/d and the temperature is degree celsius.
 
     References
     ----------
-    .. [collenteur_2020] Collenteur, R.A., Bakker, M., Klammler, G., & Birk,
-       S. (in Review) Estimating groundwater recharge from groundwater
-       levels using non-linear transfer function noise models and comparison to
-       lysimeter data. https://doi.org/10.5194/hess-2020-392
+    .. [collenteur_2021] Collenteur, R. A., Bakker, M., Klammler, G., and
+       Birk, S.: Estimation of groundwater recharge from groundwater levels
+       using nonlinear transfer function noise models and comparison to
+       lysimeter data, Hydrol. Earth Syst. Sci., 25, 2931–2949,
+       https://doi.org/10.5194/hess-25-2931-2021, 2021.
+
+    .. [kavetski_2007] Kavetski, D. and Kuczera, G. (2007). Model smoothing
+       strategies to remove microscale discontinuities and  spurious
+       secondary optima  in  objective  functions  in  hydrological
+       calibration. Water Resources Research, 43(3).
+
     """
     _name = "FlexModel"
 
-    def __init__(self):
+    def __init__(self, interception=True, snow=False, gw_uptake=False):
         check_numba()
         RechargeBase.__init__(self)
-        self.nparam = 6
+        self.snow = snow
+        self.interception = interception
+        self.gw_uptake = gw_uptake
+        self.nparam = 5
+        if self.interception:
+            self.nparam += 1
+        if self.snow:
+            self.nparam += 3
 
     def get_init_parameters(self, name="recharge"):
         parameters = DataFrame(
             columns=["initial", "pmin", "pmax", "vary", "name"])
         parameters.loc[name + "_srmax"] = (250.0, 1e-5, 1e3, True, name)
         parameters.loc[name + "_lp"] = (0.25, 1e-5, 1, False, name)
-        parameters.loc[name + "_ks"] = (100.0, 1, 1e4, True, name)
-        parameters.loc[name + "_gamma"] = (4.0, 1e-5, 50.0, True, name)
-        parameters.loc[name + "_simax"] = (2.0, 1e-5, 10.0, False, name)
+        parameters.loc[name + "_ks"] = (100.0, 1e-5, 1e4, True, name)
+        parameters.loc[name + "_gamma"] = (2.0, 1e-5, 20.0, True, name)
         parameters.loc[name + "_kv"] = (1.0, 0.25, 2.0, False, name)
+        if self.interception:
+            parameters.loc[name + "_simax"] = (2.0, 1e-5, 10.0, False, name)
+        if self.snow:
+            parameters.loc[name + "_tp"] = (0.0, -4.0, 4.0, True, name)
+            parameters.loc[name + "_tk"] = (0.0, -4.0, 4.0, True, name)

```

Comment:
Is there an option available to take `tp` and `tk` as one parameter? If I understand correctly `tp` is the snowfall melt parameter and `tk` is a parameter related to the melting point of snow on the ground. I can imagine that these parameters are usually relatively similar. In the publication of _Kavetski and Kuczera_ they also take `tp = tk` for simplicity.

Additionally, what happens when `tp < tk`, then you can have days where is no snow melt but all precipitation falls as rain (and not snow)?

Reply:
On parameter `tp`:

I'm getting really small values for the standard error (e-12).

Reply:
Good points. I just merged these two parameters into one (tt), following Kavetski and Kuczera (2007), the HBV model, and many others studies Commit will follow in a bit. Probably better to start easy and do a proper study where different snow models are tested somewhere in the future.